### ISSUE TITLE
Stream from Tapo C200 using WebRTC leads to overflow of server RAM

### ISSUE THREAD

**Issue by NormKOALA** [AlexxIT/go2rtc#398]
I have two Tapo cameras - C100 and C200, both connected to Frigate12 via go2rtc (built in version 1.2.0).
With C100 relatively no problem, but watching a main stream from C200 using WebRTC leads to overflow of server RAM in the first seconds and its freezing. When using MSE all works normal.

### Friagte config:
```yaml
go2rtc:
  rtsp:
    username: "user"
    password: "pass"
  streams:
    Hall_Camera:         # <------------ Tapo C100
      - rtsp://user:pass@192.168.1.120:554/stream1
      - "ffmpeg:Hall_Camera"
    Hall_Camera_Sub:
      - rtsp://user:pass@192.168.1.120:554/stream2
      - "ffmpeg:Hall_Camera_Sub"

    Livingroom_Camera:   # <------------ Tapo C200
      - rtsp://user:pass@192.168.1.121:554/stream1
      - "ffmpeg:Livingroom__Camera"
    Livingroom_Camera_Sub:
      - rtsp://user:pass@192.168.1.121:554/stream2
      - "ffmpeg:Livingroom_Camera_Sub"

    Facade_Camera:
      - rtsp://user:pass@192.168.1.122:554/h264Preview_01_main
      - "ffmpeg:Facade_Camera"
    Facade_Camera_Sub:
      - rtsp://user:pass@192.168.1.122:554/h264Preview_01_sub
      - "ffmpeg:Facade_Camera_Sub"
  webrtc:
    candidates:
      - 192.168.1.173:8555
      - stun:8555
```

### Friagte logs:
```
2023-04-30 13:00:53.813080833  [2023-04-30 13:00:53] watchdog.Livingroom_Camera     INFO    : Livingroom_Camera exceeded fps limit. Exiting ffmpeg...
2023-04-30 13:00:53.813956004  [2023-04-30 13:00:53] watchdog.Livingroom_Camera     INFO    : Waiting for ffmpeg to exit gracefully...
2023-04-30 13:08:05.150261805  [2023-04-30 13:02:43] watchdog.Hall_Camera           INFO    : No frames received from Hall_Camera in 20 seconds. Exiting ffmpeg...
2023-04-30 13:08:05.360272636  [2023-04-30 13:04:17] watchdog.Facade_Camera         INFO    : No frames received from Facade_Camera in 20 seconds. Exiting ffmpeg...
2023-04-30 13:08:05.360488788  [2023-04-30 13:04:34] watchdog.Hall_Camera           INFO    : Waiting for ffmpeg to exit gracefully...
2023-04-30 13:08:05.360595305  [2023-04-30 13:08:05] watchdog.Facade_Camera         INFO    : Waiting for ffmpeg to exit gracefully...
2023-04-30 13:08:05.755331943  [2023-04-30 13:08:05] frigate.video                  ERROR   : Livingroom_Camera: Unable to read frames from ffmpeg process.
2023-04-30 13:08:05.767034930  [2023-04-30 13:08:05] watchdog.Livingroom_Camera     INFO    : FFmpeg did not exit. Force killing...
2023-04-30 13:08:05.770044681  [2023-04-30 13:08:05] frigate.video                  ERROR   : Livingroom_Camera: Unable to read frames from ffmpeg process.
2023-04-30 13:08:05.770225208  [2023-04-30 13:08:05] frigate.video                  ERROR   : Livingroom_Camera: Unable to read frames from ffmpeg process.
2023-04-30 13:08:05.770364796  [2023-04-30 13:08:05] frigate.video                  ERROR   : Livingroom_Camera: Unable to read frames from ffmpeg process.
2023-04-30 13:08:05.770613644  [2023-04-30 13:08:05] frigate.video                  ERROR   : Livingroom_Camera: Unable to read frames from ffmpeg process.
2023-04-30 13:08:05.770896290  [2023-04-30 13:08:05] frigate.video                  ERROR   : Livingroom_Camera: Unable to read frames from ffmpeg process.
2023-04-30 13:08:05.771266850  [2023-04-30 13:08:05] frigate.video                  ERROR   : Livingroom_Camera: Unable to read frames from ffmpeg process.
2023-04-30 13:08:05.771394431  [2023-04-30 13:08:05] frigate.video                  ERROR   : Livingroom_Camera: ffmpeg process is not running. exiting capture thread...
2023-04-30 13:08:06.046025547  [2023-04-30 13:08:06] frigate.video                  ERROR   : Facade_Camera: Unable to read frames from ffmpeg process.
2023-04-30 13:08:06.046960003  [2023-04-30 13:08:06] frigate.video                  ERROR   : Facade_Camera: Unable to read frames from ffmpeg process.
2023-04-30 13:08:06.047139017  [2023-04-30 13:08:06] frigate.video                  ERROR   : Facade_Camera: ffmpeg process is not running. exiting capture thread...
2023-04-30 13:08:06.050627019  [2023-04-30 13:08:06] frigate.video                  ERROR   : Hall_Camera: Unable to read frames from ffmpeg process.
2023-04-30 13:08:06.051186416  [2023-04-30 13:08:06] frigate.video                  ERROR   : Hall_Camera: ffmpeg process is not running. exiting capture thread...
2023-04-30 13:08:06.219069864  [2023-04-30 13:08:06] ffmpeg.Facade_Camera.record    ERROR   : [segment @ 0x55ec33240c40] Timestamps are unset in a packet for stream 0. This is deprecated and will stop working in the future. Fix your code to set the timestamps properly
2023-04-30 13:08:06.219378387  [2023-04-30 13:08:06] ffmpeg.Facade_Camera.record    ERROR   : [segment @ 0x563071da4c40] Timestamps are unset in a packet for stream 0. This is deprecated and will stop working in the future. Fix your code to set the timestamps properly
2023-04-30 13:08:06.219537546  [2023-04-30 13:08:06] watchdog.Facade_Camera         INFO    : Terminating the existing ffmpeg process...
2023-04-30 13:08:06.219768719  [2023-04-30 13:08:06] watchdog.Facade_Camera         INFO    : Waiting for ffmpeg to exit gracefully...
2023-04-30 13:08:15.771080804  [2023-04-30 13:08:15] watchdog.Livingroom_Camera     ERROR   : Ffmpeg process crashed unexpectedly for Livingroom_Camera.
2023-04-30 13:08:15.775666129  [2023-04-30 13:08:15] watchdog.Livingroom_Camera     ERROR   : The following ffmpeg logs include the last 100 lines prior to exit.
2023-04-30 13:08:15.775749779  [2023-04-30 13:08:15] ffmpeg.Livingroom_Camera.detect ERROR   : [rtsp @ 0x559d8689bd40] RTP: PT=60: bad cseq d5ec expected=470c
2023-04-30 13:08:15.776281756  [2023-04-30 13:08:15] ffmpeg.Livingroom_Camera.detect ERROR   : [out_0_0 @ 0x559d87252e80] 100 buffers queued in out_0_0, something may be wrong.
2023-04-30 13:08:15.776287857  [2023-04-30 13:08:15] ffmpeg.Livingroom_Camera.detect ERROR   : [out_0_0 @ 0x559d87252e80] 1000 buffers queued in out_0_0, something may be wrong.
2023-04-30 13:08:16.051908859  [2023-04-30 13:08:16] watchdog.Hall_Camera           ERROR   : Ffmpeg process crashed unexpectedly for Hall_Camera.
2023-04-30 13:08:16.052169779  [2023-04-30 13:08:16] watchdog.Hall_Camera           ERROR   : The following ffmpeg logs include the last 100 lines prior to exit.
2023-04-30 13:08:16.232091760  [2023-04-30 13:08:16] watchdog.Facade_Camera         ERROR   : Ffmpeg process crashed unexpectedly for Facade_Camera.
2023-04-30 13:08:16.232243674  [2023-04-30 13:08:16] watchdog.Facade_Camera         ERROR   : The following ffmpeg logs include the last 100 lines prior to exit.
```

### go2rtc logs:
```
2023-04-30 12:47:01.337354441  12:47:01.337 ERR github.com/AlexxIT/go2rtc/cmd/streams/producer.go:53 > error="exec: exit status 1" url=ffmpeg:Livingroom__Camera
2023-04-30 12:54:28.380555574  12:54:28.380 ERR github.com/AlexxIT/go2rtc/cmd/streams/producer.go:53 > error="exec: exit status 1" url=ffmpeg:Livingroom__Camera
2023-04-30 12:54:28.418204405  12:54:28.417 ERR github.com/AlexxIT/go2rtc/cmd/streams/producer.go:53 > error="exec: exit status 1" url=ffmpeg:Livingroom__Camera
2023-04-30 12:58:01.703655466  2023/04/30 12:58:01 [INFO] mdns: Closing client {true false 0xc00030e0a8 <nil> 0xc00030e0b8 <nil> 1 0xc0000200c0}
2023-04-30 13:01:39.116301471  13:01:39.090 WRN github.com/AlexxIT/go2rtc/cmd/streams/producer.go:133 > error="read tcp 172.18.0.2:35252->192.168.1.122:554: i/o timeout" url=rtsp://user:pass@192.168.1.122:554/h264Preview_01_sub
2023-04-30 13:08:15.826868373  13:08:15.826 ERR github.com/AlexxIT/go2rtc/cmd/streams/producer.go:53 > error="exec: exit status 1" url=ffmpeg:Livingroom__Camera
2023-04-30 13:10:37.121814862  13:10:37.121 ERR github.com/AlexxIT/go2rtc/cmd/streams/producer.go:53 > error="exec: exit status 1" url=ffmpeg:Livingroom__Camera
```

**Comment by AlexxIT** [comment]
You have very strange config. Every your camera has FFmpeg loopback source. That do nothing useful but can be source of a trouble

**Comment by NormKOALA** [comment]
Could you tell me a little more about what is strange.
I did everything according to the Frigate documentation. Like this:

<img width="728" alt="Screenshot 2023-04-30 162153" src="https://user-images.githubusercontent.com/70369365/235341004-edb29836-d845-4d70-96e7-cdac4e356fd9.png">

### Whole my config:
```yaml
mqtt:
  host: 192.168.1.173
  user: user
  password: pass

ffmpeg:
  hwaccel_args: preset-nvidia-h264

go2rtc:
  rtsp:
    username: "user"
    password: "pass"
  streams:
    Hall_Camera:
      - rtsp://user:pass@192.168.1.120:554/stream1
      - "ffmpeg:Hall_Camera"
    Hall_Camera_Sub:
      - rtsp://user:pass@192.168.1.120:554/stream2
      - "ffmpeg:Hall_Camera_Sub"

    Livingroom_Camera:
      - rtsp://user:pass@192.168.1.121:554/stream1
      - "ffmpeg:Livingroom__Camera"
    Livingroom_Camera_Sub:
      - rtsp://user:pass@192.168.1.121:554/stream2
      - "ffmpeg:Livingroom_Camera_Sub"

    Facade_Camera:
      - rtsp://user:pass@192.168.1.122:554/h264Preview_01_main
      - "ffmpeg:Facade_Camera"
    Facade_Camera_Sub:
      - rtsp://user:pass@192.168.1.122:554/h264Preview_01_sub
      - "ffmpeg:Facade_Camera_Sub"
  webrtc:
    candidates:
      - 192.168.1.173:8555
      - stun:8555

cameras:
  Hall_Camera:
    ffmpeg:
      output_args:
        record: preset-record-generic-audio-copy
      inputs:
        # - path: rtsp://127.0.0.1:8554/Hall_Camera_Sub
        #   input_args: preset-rtsp-restream
        #   roles:
        #   - detect
        - path: rtsp://127.0.0.1:8554/Hall_Camera
          input_args: preset-rtsp-restream
          roles:
          - detect
          - record
    detect:
      enabled: true
      max_disappeared: 25
      width: 1920
      height: 1080
      fps: 5
    objects:
      track:
      - person
    snapshots:
      enabled: true
      bounding_box: true
      retain:
        default: 2 # days
    record:
      enabled: true
      events:
        retain:
          default: 2 # days
    motion:
      mask:
      - 788,0,808,398,1015,391,1020,0,1920,0,1920,1080,0,1080,0,0
    live:
      stream_name: Hall_Camera_Sub


  Livingroom_Camera:
    ffmpeg:
      output_args:
        record: preset-record-generic-audio-copy
      inputs:
        - path: rtsp://127.0.0.1:8554/Livingroom_Camera
          input_args: preset-rtsp-restream
          roles:
          - detect
    detect:
      enabled: false
      max_disappeared: 25
      width: 1920
      height: 1080
      fps: 5
    objects:
      track:
      - person
    live:
      stream_name: Livingroom_Camera_Sub


  Facade_Camera:
    ffmpeg:
      output_args:
        record: preset-record-generic-audio-copy
      inputs:
        - path: rtsp://127.0.0.1:8554/Facade_Camera_Sub
          input_args: preset-rtsp-restream
          roles:
          - detect
        - path: rtsp://127.0.0.1:8554/Facade_Camera
          input_args: preset-rtsp-restream
          roles:
          - record
    detect:
      enabled: true
      max_disappeared: 25
      width: 2560
      height: 1920
      fps: 5
    objects:
      track:
      - person
    snapshots:
      enabled: true
      bounding_box: true
      retain:
        default: 2 # days
    record:
      enabled: true
      events:
        retain:
          default: 2 # days
    motion:
      mask:
      - 819,272,1377,268,2111,314,2560,750,2560,0,0,0,0,386,251,347
    live:
      stream_name: Facade_Camera_Sub

detectors:
  coral:
    type: edgetpu
    device: usb

logger:
  logs:
    frigate.edgetpu: debug

```

Looks the same to me.

**Comment by AlexxIT** [comment]
Your config not the same. Check carefully. You need to use ffmpeg source only when you "need" it.
https://github.com/AlexxIT/go2rtc#source-ffmpeg

**Comment by NormKOALA** [comment]
I still can not understand.

I removed rows:
```yaml
    live:
      stream_name:
```
and added `#audio=opus` in rows like `- "ffmpeg:Hall_Camera"`

After that, my system crashes by all streams using WebRTC.
It's very inconvenient restart system every time.
Could you just highlight problem place for me, without quizzes. Thank you.

**Comment by Uriziel01** [comment]
TL;DR; - @NormKOALA  What @AlexxIT meant is that you do not need to add ffmpeg entries like `- "ffmpeg:Hall_Camera` all over the place if you are not doing any actual video/audio recoding, just remove/comment out all the lines starting with `ffmpeg`

As you can see in https://docs.frigate.video/guides/configuring_go2rtc/#what-if-my-video-doesnt-play the ffmpeg entry is used only when you want to do some specific ffmpeg transcoding when having issues with compatibility and so on.

I have Tapo C200 as well as Tapo C210 in my current setup and both of them are working fine without any memory issues, so let's try to get rid of the ffmpeg loopback sources as @AlexxIT mentioned and then let's validate if anything changed.

**Comment by Uriziel01** [comment]
And you do not need to add the `#audio=opus` to C200/C210 stream just to watch it in the browser/android client, it works fine without any reencoding. Not sure which one are you using ATM.

**Comment by NormKOALA** [comment]
@Uriziel01 Thank you for trying to help.

>TL;DR; - @NormKOALA What @AlexxIT meant is that you do not need to add ffmpeg entries like - "ffmpeg:Hall_Camera all over the place if you are not doing any actual video/audio recoding, just remove/comment out all the lines starting with ffmpeg

I removed all lines starting with `ffmpeg` and it didn't help.

>As you can see in https://docs.frigate.video/guides/configuring_go2rtc/#what-if-my-video-doesnt-play the ffmpeg entry is used only when you want to do some specific ffmpeg transcoding when having issues with compatibility and so on.

I understand it, but the [Next step](https://docs.frigate.video/configuration/restream#reduce-connections-to-camera) says that in order to perform `Reduce Connections To Camera`, you need to add the line like `ffmpeg:Hall_Camera#audio=opus`, (_probably not, but it's not obvious from the Frigarte documentation_)

>I have Tapo C200 as well as Tapo C210 in my current setup and both of them are working fine without any memory issues, so let's try to get rid of the ffmpeg loopback sources as @AlexxIT mentioned and then let's validate if anything changed.

My current config:

```yaml
mqtt:
  host: 192.168.1.173
  user: user
  password: pass

ffmpeg:
  hwaccel_args: preset-nvidia-h264

go2rtc:
  rtsp:
    username: "user"
    password: "pass"
  streams:
    Hall_Camera:
      - rtsp://user:pass@192.168.1.120:554/stream1
      # - "ffmpeg:Hall_Camera#audio=opus"
    Hall_Camera_Sub:
      - rtsp://user:pass@192.168.1.120:554/stream2
      # - "ffmpeg:Hall_Camera_Sub#audio=opus"

    Livingroom_Camera:
      - rtsp://user:pass@192.168.1.121:554/stream1
      # - "ffmpeg:Livingroom_Camera#audio=opus"
    Livingroom_Camera_Sub:
      - rtsp://user:pass@192.168.1.121:554/stream2
      # - "ffmpeg:Livingroom_Camera_Sub#audio=opus"

    Facade_Camera:
      - rtsp://user:pass@192.168.1.122:554/h264Preview_01_main
      # - "ffmpeg:Facade_Camera#audio=opus"
    Facade_Camera_Sub:
      - rtsp://user:pass@192.168.1.122:554/h264Preview_01_sub
      # - "ffmpeg:Facade_Camera_Sub#audio=opus"
  webrtc:
    candidates:
      - 192.168.1.173:8555
      - stun:8555

cameras:
  Hall_Camera:
    ffmpeg:
      # output_args:
      #   record: preset-record-generic-audio-copy
      inputs:
        # - path: rtsp://127.0.0.1:8554/Hall_Camera_Sub
        #   input_args: preset-rtsp-restream
        #   roles:
        #   - detect
        - path: rtsp://127.0.0.1:8554/Hall_Camera
          input_args: preset-rtsp-restream
          roles:
          - detect
          - record
    detect:
      enabled: true
      max_disappeared: 25
      width: 1920
      height: 1080
      fps: 5
    objects:
      track:
      - person
    snapshots:
      enabled: true
      bounding_box: true
      retain:
        default: 2 # days
    record:
      enabled: true
      events:
        retain:
          default: 2 # days
    motion:
      mask:
      - 788,0,808,398,1015,391,1020,0,1920,0,1920,1080,0,1080,0,0
    live:
      stream_name: Hall_Camera_Sub


  Livingroom_Camera:
    ffmpeg:
      # output_args:
      #   record: preset-record-generic-audio-copy
      inputs:
        - path: rtsp://127.0.0.1:8554/Livingroom_Camera
          input_args: preset-rtsp-restream
          roles:
          - detect
    detect:
      enabled: false
      max_disappeared: 25
      width: 1920
      height: 1080
      fps: 5
    objects:
      track:
      - person
    live:
      stream_name: Livingroom_Camera_Sub


  Facade_Camera:
    ffmpeg:
      # output_args:
      #   record: preset-record-generic-audio-copy
      inputs:
        - path: rtsp://127.0.0.1:8554/Facade_Camera_Sub
          input_args: preset-rtsp-restream
          roles:
          - detect
        - path: rtsp://127.0.0.1:8554/Facade_Camera
          input_args: preset-rtsp-restream
          roles:
          - record
    detect:
      enabled: true
      max_disappeared: 25
      width: 2560
      height: 1920
      fps: 5
    objects:
      track:
      - person
    snapshots:
      enabled: true
      bounding_box: true
      retain:
        default: 2 # days
    record:
      enabled: true
      events:
        retain:
          default: 2 # days
    motion:
      mask:
      - 819,272,1377,268,2111,314,2560,750,2560,0,0,0,0,386,251,347
    live:
      stream_name: Facade_Camera_Sub

detectors:
  coral:
    type: edgetpu
    device: usb

logger:
  logs:
    frigate.edgetpu: debug
```

With these settings, the problem began to appear randomly. Now system can crash when Frigate detected person and try to record clip.

**Comment by NickM-27** [comment]
> With these settings, the problem began to appear randomly. Now system can crash when Frigate detected person and try to record clip.

that's not really accurate. Frigate is always recording, it just discards the recording segments which don't match the retention config.

**Comment by NormKOALA** [comment]
@NickM-27 Ok.

Now system can crash when Frigate detected person ~and try to record clip~.

**Comment by Write** [comment]
For what is worth, I think it's a specific issue when ffmpeg crash.

I relied upon an external rtsp stream from within go2rtc, if this external rtsp stream wasn't closed properly, go2rtc's would fill up the whole ram and crash the whole system in a matter of seconds. 

I had the exact same issue when setting up rtsp-media-server (now mediamtx) as I wanted to set it as a systemctl service, I had specifically add KillSignal=SIGINT, otherwise ffmpeg sub-process of mediamtx would simply fill up the whole RAM when I ran the command "systemctl restart mediamtx".

I find that sometimes the same issue arise with go2rtc when streams are not closed properly / too abruptly / wrong signal used to kill it. (Really not sure).

And I'm pretty sure your first issue filling the whole RAM was kinda the same bug as me, where there was an issue with an ffmpeg process.

**Comment by AlexxIT** [comment]
Please let me know if the problem is relevant
